moshi/src/main/java/com/squareup/moshi/JsonValueWriter.kt

```diff
@@ -0,0 +1,265 @@
+    check(peekScope() == EMPTY_ARRAY) { "Nesting problem." }
+    if (stackSize == flattenStackSize.inv()) {
+      // Cancel this close. Restore the flattenStackSize so we're ready to flatten again!
+      flattenStackSize = flattenStackSize.inv()
+      return this
+    }
+    stackSize--
+    stack[stackSize] = null
+    pathIndices[stackSize - 1]++
+    return this
+  }
+
+  override fun beginObject(): JsonWriter {
+    check(!promoteValueToName) { "Object cannot be used as a map key in JSON at path $path" }
+    if (stackSize == flattenStackSize && scopes[stackSize - 1] == EMPTY_OBJECT) {
+      // Cancel this open. Invert the flatten stack size until this is closed.
+      flattenStackSize = flattenStackSize.inv()
+      return this
+    }
+    checkStack()
+    val map = LinkedHashTreeMap<String, Any>()
+    add(map)
+    stack[stackSize] = map
+    pushScope(EMPTY_OBJECT)
+    return this
+  }
+
+  override fun endObject(): JsonWriter {
+    check(peekScope() == EMPTY_OBJECT) { "Nesting problem." }
+    check(deferredName == null) { "Dangling name: $deferredName" }
+    if (stackSize == flattenStackSize.inv()) {
+      // Cancel this close. Restore the flattenStackSize so we're ready to flatten again!
+      flattenStackSize = flattenStackSize.inv()
+      return this
+    }
+    promoteValueToName = false
+    stackSize--
+    stack[stackSize] = null
+    pathNames[stackSize] = null // Free the last path name so that it can be garbage collected!
+    pathIndices[stackSize - 1]++
+    return this
+  }
+
+  override fun name(name: String): JsonWriter {
+    check(stackSize != 0) { "JsonWriter is closed." }
+    check(peekScope() == EMPTY_OBJECT && deferredName == null && !promoteValueToName) { "Nesting problem." }
+    deferredName = name
+    pathNames[stackSize - 1] = name
+    return this
+  }
+
+  override fun value(value: String?): JsonWriter {
+    if (promoteValueToName) {
+      promoteValueToName = false
+      return name(value!!)
+    }
+    add(value)
+    pathIndices[stackSize - 1]++
+    return this
+  }
+
+  override fun nullValue(): JsonWriter {
+    check(!promoteValueToName) { "null cannot be used as a map key in JSON at path $path" }
+    add(null)
+    pathIndices[stackSize - 1]++
+    return this
+  }
+
+  override fun value(value: Boolean): JsonWriter {
+    check(!promoteValueToName) { "Boolean cannot be used as a map key in JSON at path $path" }
+    add(value)
+    pathIndices[stackSize - 1]++
+    return this
+  }
+
+  override fun value(value: Boolean?): JsonWriter {
+    check(!promoteValueToName) { "Boolean cannot be used as a map key in JSON at path $path" }
+    add(value)
+    pathIndices[stackSize - 1]++
+    return this
+  }
+
+  override fun value(value: Double): JsonWriter {
+    require(
+      !(
+        !lenient &&
+          (java.lang.Double.isNaN(value) || value == Double.NEGATIVE_INFINITY || value == Double.POSITIVE_INFINITY)
+        )
+    ) { "Numeric values must be finite, but was $value" }
```

Comment:
can the `!` be propagated so this can be simplified?

Reply:
oof yeah let me clean this up

Reply:
21b6dc4